macaw-dump/src/Data/Macaw/Dump/Memory.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Data.Macaw.Dump.Memory
+  ( MemoryConfig(..)
+  , memoryConfig
+  , memory
+  ) where
+
+import Data.ElfEdit qualified as EE
+import Data.Macaw.Architecture.Info qualified as MAI
+import Data.Macaw.BinaryLoader qualified as Loader
+import Data.Macaw.CFG.Core qualified as MC
+import Data.Macaw.Dump.CLIUtils qualified as MDCU
+import Data.Macaw.Memory.LoadCommon qualified as LC
+import Data.Macaw.Memory qualified as MM
+import Options.Applicative qualified as Opt
+
+data MemoryConfig
+  = MemoryConfig
+    { memBinPath :: FilePath
+    }
+
+memoryConfig :: Opt.Parser MemoryConfig
+memoryConfig =
+  MemoryConfig
+  <$> MDCU.binOpt
+
+-- Currently, we do not apply any offsets to addresses in the loaded binary. We
+-- will need to reconsider this if we want to support shared libraries.
+loadOptions :: LC.LoadOptions
+loadOptions =
+  LC.LoadOptions
+    { LC.loadOffset = Nothing
+    }
+
+memory ::
+  forall arch.
+  ( MM.MemWidth (MC.ArchAddrWidth arch)
+  , Loader.BinaryLoader arch (EE.ElfHeaderInfo (MC.ArchAddrWidth arch))
+  ) =>
+  MAI.ArchitectureInfo arch ->
+  MemoryConfig ->
+  IO ()
+memory archInfo cfg = do
+  ehi <- MDCU.loadElf archInfo (memBinPath cfg)
+  loaded <- Loader.loadBinary @arch loadOptions ehi
+  let mem = Loader.memoryImage loaded
+  print mem
```

Comment:
I wonder if we should also consider printing the [`relativeSegmentContents`](https://github.com/GaloisInc/macaw/blob/aedf20208aacfebadc901519b28a675480f9ded7/base/src/Data/Macaw/Memory.hs#L1641-L1649) of the `Memory`, as that gives finer-grained insight than you would get if you only looked at the `Show` output for `Memory` (which simply uses `memSegments`). Moreover, `macaw-symbolic` creates its model of the global address space using `relativeSegmentContents`, so it's pretty pertinent information.

Reply:
Yeah, perhaps so? Honestly I find that function's haddock and implementation pretty opaque...

Reply:
My high-level understanding is that it just tells you which parts of the `Memory`'s segments have actual contents (bytes, BSS, or relocations) backing it. For instance, even if you have a segment ranging from `[0x0, 0x1000)`, there's no guarantee that all addresses actually point to useful contents—that information can only be gleaned by looking at the `relativeSegmentContents`.

Reply:
Hmm, I find that confusing. The docs for `MemorySegment` say that they represent "a contiguous sequence of bytes in memory".

https://github.com/GaloisInc/macaw/blob/aedf20208aacfebadc901519b28a675480f9ded7/base/src/Data/Macaw/Memory.hs#L808

This is confusing, because `SegmentContents` is a *map*, whereas a continguous sequence would perhaps be more naturall represented as a list/vector:

https://github.com/GaloisInc/macaw/blob/aedf20208aacfebadc901519b28a675480f9ded7/base/src/Data/Macaw/Memory.hs#L612

Perhaps we should stop using the word "contiguous" to describe a `MemorySegment`?

What is at the addresses that are not bytes, zeroes, or relocations? Are those addresses unmapped?

Reply:
Oh, perhaps my understanding is off then. Indeed, looking at a small example, it does appear that the `LOAD` headers of the binary do pretty tightly correspond with the segments.

In that case, printing the contents would only be useful if you want to know what is contained in each segment.

Reply:
To be clear, I'm not disagreeing with anything you're saying or suggesting - I just actually don't understand whether or not the `MemChunk`s in a `MemSegment` are supposed to cover the whole range of addresses within the segment (i.e., be contiguous) or not.

Reply:
In lieu of evidence to the contrary, I'll assume that they are in fact suppose to cover the whole range.

Reply:
Well, this is exactly the sort of question that I hope `macaw-dump` can help us answer, so I've gone ahead and implemented printing the content of memory as suggested (off by default, available via `--contents`), and it does look to me that at least in simple cases the `MemChunk`s are adjacent.